core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -584,16 +584,21 @@ public void runInContext() {
 
     @Override
     public void closed(final Status status) {
+      // For cancellations, promptly inform any users of the context that their work should be
+      // aborted. Otherwise, we can wait until pending work is done.
+      if (!status.isOk()) {
+        context.cancel(status.getCause());
```

Comment:
The callExecutor is the one doing the user supplied unit of work, the same one that is should be checking for `Context.current().isCancelled()` or `ServerCall.isCancelled()`. We want to be able to set the flag and have its effect seen by the callExecutor's runnable, rather than enqueuing it into callExecutor.